[deutschebahn] DeutscheBahnTimetableHandlerTest unstable

This unit test failed in https://github.com/openhab/openhab-addons/actions/runs/1805107390 :

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.708 s <<< FAILURE! - in org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest
[ERROR] org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest.testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided  Time elapsed: 2.056 s  <<< FAILURE!
org.mockito.exceptions.verification.WantedButNotInvoked: 

Wanted but not invoked:
thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:arrival#line,
    UNDEF
);
-> at org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest.verifyChannelsUpdatedToUndef(DeutscheBahnTimetableHandlerTest.java:150)

However, there were exactly 4 interactions with this mock:
thingHandlerCallback.statusUpdated(
    Mock for Bridge, hashCode: 1414934182,
    UNKNOWN
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:344)

thingHandlerCallback.statusUpdated(
    Mock for Bridge, hashCode: 1414934182,
    ONLINE
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:344)

thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:trip#category,
    UNDEF
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:232)

thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:arrival#planned-time,
    UNDEF
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:232)


	at org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest.verifyChannelsUpdatedToUndef(DeutscheBahnTimetableHandlerTest.java:150)
	at org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest.testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided(DeutscheBahnTimetableHandlerTest.java:180)

[INFO] Running org.openhab.binding.deutschebahn.internal.EventAttributeTest
[INFO] Tests run: 23, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.183 s - in org.openhab.binding.deutschebahn.internal.EventAttributeTest
[INFO] Running org.openhab.binding.deutschebahn.internal.DeutscheBahnTrainHandlerTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.094 s - in org.openhab.binding.deutschebahn.internal.DeutscheBahnTrainHandlerTest
[INFO] Running org.openhab.binding.deutschebahn.internal.filter.TimetableByStringEventAttributeFilterTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.015 s - in org.openhab.binding.deutschebahn.internal.filter.TimetableByStringEventAttributeFilterTest
[INFO] Running org.openhab.binding.deutschebahn.internal.filter.FilterScannerTest
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.024 s - in org.openhab.binding.deutschebahn.internal.filter.FilterScannerTest
[INFO] Running org.openhab.binding.deutschebahn.internal.filter.FilterParserTest
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.022 s - in org.openhab.binding.deutschebahn.internal.filter.FilterParserTest
[INFO] Running org.openhab.binding.deutschebahn.internal.timetable.TimetableLoaderTest
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.373 s - in org.openhab.binding.deutschebahn.internal.timetable.TimetableLoaderTest
[INFO] Running org.openhab.binding.deutschebahn.internal.timetable.TimetablesV1ImplTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.331 s - in org.openhab.binding.deutschebahn.internal.timetable.TimetablesV1ImplTest
[INFO] Running org.openhab.binding.deutschebahn.internal.TripLabelAttributeTest
[INFO] Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.026 s - in org.openhab.binding.deutschebahn.internal.TripLabelAttributeTest
[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DeutscheBahnTimetableHandlerTest.testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided:180->verifyChannelsUpdatedToUndef:150 
Wanted but not invoked:
thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:arrival#line,
    UNDEF
);
-> at org.openhab.binding.deutschebahn.internal.DeutscheBahnTimetableHandlerTest.verifyChannelsUpdatedToUndef(DeutscheBahnTimetableHandlerTest.java:150)

However, there were exactly 4 interactions with this mock:
thingHandlerCallback.statusUpdated(
    Mock for Bridge, hashCode: 1414934182,
    UNKNOWN
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:344)

thingHandlerCallback.statusUpdated(
    Mock for Bridge, hashCode: 1414934182,
    ONLINE
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateStatus(BaseThingHandler.java:344)

thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:trip#category,
    UNDEF
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:232)

thingHandlerCallback.stateUpdated(
    deutschebahn:train:train-2:arrival#planned-time,
    UNDEF
);
-> at org.openhab.core.thing.binding.BaseThingHandler.updateState(BaseThingHandler.java:232)


[INFO] 
[ERROR] Tests run: 57, Failures: 1, Errors: 0, Skipped: 0

```

Can you have a look at it @soenkekueper?

--- a/bundles/org.openhab.binding.deutschebahn/src/main/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnHandlerFactory.java
+++ b/bundles/org.openhab.binding.deutschebahn/src/main/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnHandlerFactory.java
@@ -50,7 +50,7 @@ public class DeutscheBahnHandlerFactory extends BaseThingHandlerFactory {
         final ThingTypeUID thingTypeUID = thing.getThingTypeUID();
 
         if (TIMETABLE_TYPE.equals(thingTypeUID)) {
-            return new DeutscheBahnTimetableHandler((Bridge) thing, TimetablesV1Impl::new, Date::new);
+            return new DeutscheBahnTimetableHandler((Bridge) thing, TimetablesV1Impl::new, Date::new, null);
         } else if (TRAIN_TYPE.equals(thingTypeUID)) {
             return new DeutscheBahnTrainHandler(thing);
         }

--- a/bundles/org.openhab.binding.deutschebahn/src/main/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandler.java
+++ b/bundles/org.openhab.binding.deutschebahn/src/main/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandler.java
@@ -20,6 +20,7 @@ import java.util.Date;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.locks.Lock;
@@ -108,9 +109,11 @@ public class DeutscheBahnTimetableHandler extends BaseBridgeHandler {
     private final Logger logger = LoggerFactory.getLogger(DeutscheBahnTimetableHandler.class);
     private @Nullable TimetableLoader loader;
 
-    private TimetablesV1ApiFactory timetablesV1ApiFactory;
+    private final TimetablesV1ApiFactory timetablesV1ApiFactory;
 
-    private Supplier<Date> currentTimeProvider;
+    private final Supplier<Date> currentTimeProvider;
+
+    private final ScheduledExecutorService executorService;
 
     /**
      * Creates an new {@link DeutscheBahnTimetableHandler}.
@@ -118,10 +121,12 @@ public class DeutscheBahnTimetableHandler extends BaseBridgeHandler {
     public DeutscheBahnTimetableHandler( //
             final Bridge bridge, //
             final TimetablesV1ApiFactory timetablesV1ApiFactory, //
-            final Supplier<Date> currentTimeProvider) {
+            final Supplier<Date> currentTimeProvider, //
+            @Nullable final ScheduledExecutorService executorService) {
         super(bridge);
         this.timetablesV1ApiFactory = timetablesV1ApiFactory;
         this.currentTimeProvider = currentTimeProvider;
+        this.executorService = executorService == null ? this.scheduler : executorService;
     }
 
     private List<TimetableStop> loadTimetable() {
@@ -178,7 +183,7 @@ public class DeutscheBahnTimetableHandler extends BaseBridgeHandler {
 
             this.updateStatus(ThingStatus.UNKNOWN);
 
-            this.scheduler.execute(() -> {
+            this.executorService.execute(() -> {
                 this.updateChannels();
                 this.restartJob();
             });
@@ -204,7 +209,7 @@ public class DeutscheBahnTimetableHandler extends BaseBridgeHandler {
         try {
             this.stopUpdateJob();
             if (this.getThing().getStatus() == ThingStatus.ONLINE) {
-                this.updateJob = this.scheduler.scheduleWithFixedDelay(//
+                this.updateJob = this.executorService.scheduleWithFixedDelay(//
                         this::updateChannels, //
                         0L, //
                         UPDATE_INTERVAL_SECONDS, //

--- a/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
+++ b/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
@@ -40,7 +40,6 @@ import org.openhab.core.thing.Thing;
 import org.openhab.core.thing.ThingStatus;
 import org.openhab.core.thing.ThingUID;
 import org.openhab.core.thing.binding.ThingHandlerCallback;
-import org.openhab.core.types.State;
 import org.openhab.core.types.UnDefType;
 
 /**
@@ -161,7 +160,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
     }
 
     @Test
-    public void testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided()  {
+    public void testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided() {
         final Bridge bridge = mockBridge();
         final ThingHandlerCallback callback = mock(ThingHandlerCallback.class);
 

--- a/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
+++ b/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
@@ -110,8 +110,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
 
         try {
             verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.UNKNOWN)));
-            verify(callback, timeout(1000)).statusUpdated(eq(bridge),
-                    argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
+            verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
 
             verifyThingUpdated(bridge, 0, "-5296516961807204721-2108160906-5");
             verifyThingUpdated(bridge, 1, "-8364795265993682073-2108160911-6");
@@ -124,8 +123,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
     private void verifyThingUpdated(final Bridge bridge, int offset, String stopId) {
         final Thing train = bridge.getThings().get(offset);
         final DeutscheBahnTrainHandler childHandler = (DeutscheBahnTrainHandler) train.getHandler();
-        verify(childHandler, timeout(1000))
-                .updateChannels(argThat((TimetableStop stop) -> stop.getId().equals(stopId)));
+        verify(childHandler).updateChannels(argThat((TimetableStop stop) -> stop.getId().equals(stopId)));
     }
 
     @Test
@@ -140,8 +138,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
 
         try {
             verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.UNKNOWN)));
-            verify(callback, timeout(1000)).statusUpdated(eq(bridge),
-                    argThat(arg -> arg.getStatus().equals(ThingStatus.OFFLINE)));
+            verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.OFFLINE)));
 
             verifyChannelsUpdatedToUndef(bridge, 0, callback);
             verifyChannelsUpdatedToUndef(bridge, 1, callback);
@@ -181,8 +178,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
 
         try {
             verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.UNKNOWN)));
-            verify(callback, timeout(1000)).statusUpdated(eq(bridge),
-                    argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
+            verify(callback).statusUpdated(eq(bridge), argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
 
             verifyThingUpdated(bridge, 0, stop01.getId());
             verifyChannelsUpdatedToUndef(bridge, 1, callback);

--- a/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
+++ b/bundles/org.openhab.binding.deutschebahn/src/test/java/org/openhab/binding/deutschebahn/internal/DeutscheBahnTimetableHandlerTest.java
@@ -19,10 +19,12 @@ import java.util.ArrayList;
 import java.util.Calendar;
 import java.util.GregorianCalendar;
 import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
-import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+import org.mockito.invocation.InvocationOnMock;
 import org.openhab.binding.deutschebahn.internal.timetable.TimeproviderStub;
 import org.openhab.binding.deutschebahn.internal.timetable.TimetablesV1ApiFactory;
 import org.openhab.binding.deutschebahn.internal.timetable.TimetablesV1ApiStub;
@@ -47,7 +49,6 @@ import org.openhab.core.types.UnDefType;
  * @author Sönke Küper - initial contribution.
  */
 @NonNullByDefault
-@Disabled("https://github.com/openhab/openhab-addons/issues/12235")
 public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHelper {
 
     private static Configuration createConfig() {
@@ -84,11 +85,18 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
     private DeutscheBahnTimetableHandler createAndInitHandler( //
             final ThingHandlerCallback callback, //
             final Bridge bridge, //
-            final TimetablesV1ApiFactory apiFactory) throws Exception { //
+            final TimetablesV1ApiFactory apiFactory) { //
         final TimeproviderStub timeProvider = new TimeproviderStub();
         timeProvider.time = new GregorianCalendar(2021, Calendar.AUGUST, 16, 9, 30);
 
-        final DeutscheBahnTimetableHandler handler = new DeutscheBahnTimetableHandler(bridge, apiFactory, timeProvider);
+        final ScheduledExecutorService executorStub = Mockito.mock(ScheduledExecutorService.class);
+        doAnswer((InvocationOnMock invocation) -> {
+            ((Runnable) invocation.getArguments()[0]).run();
+            return null;
+        }).when(executorStub).execute(any(Runnable.class));
+
+        final DeutscheBahnTimetableHandler handler = new DeutscheBahnTimetableHandler(bridge, apiFactory, timeProvider,
+                executorStub);
         handler.setCallback(callback);
         handler.initialize();
         return handler;
@@ -122,7 +130,7 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
     }
 
     @Test
-    public void testUpdateTrainsToUndefinedIfNoDataWasProvided() throws Exception {
+    public void testUpdateTrainsToUndefinedIfNoDataWasProvided() {
         final Bridge bridge = mockBridge();
         final ThingHandlerCallback callback = mock(ThingHandlerCallback.class);
 
@@ -136,25 +144,24 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
             verify(callback, timeout(1000)).statusUpdated(eq(bridge),
                     argThat(arg -> arg.getStatus().equals(ThingStatus.OFFLINE)));
 
-            verifyChannelsUpdatedToUndef(bridge, 0, callback, UnDefType.UNDEF);
-            verifyChannelsUpdatedToUndef(bridge, 1, callback, UnDefType.UNDEF);
-            verifyChannelsUpdatedToUndef(bridge, 2, callback, UnDefType.UNDEF);
+            verifyChannelsUpdatedToUndef(bridge, 0, callback);
+            verifyChannelsUpdatedToUndef(bridge, 1, callback);
+            verifyChannelsUpdatedToUndef(bridge, 2, callback);
 
         } finally {
             handler.dispose();
         }
     }
 
-    private static void verifyChannelsUpdatedToUndef(Bridge bridge, int offset, ThingHandlerCallback callback,
-            State expectedState) {
+    private static void verifyChannelsUpdatedToUndef(Bridge bridge, int offset, ThingHandlerCallback callback) {
         final Thing thing = bridge.getThings().get(offset);
         for (Channel channel : thing.getChannels()) {
-            verify(callback).stateUpdated(eq(channel.getUID()), eq(expectedState));
+            verify(callback).stateUpdated(eq(channel.getUID()), eq(UnDefType.UNDEF));
         }
     }
 
     @Test
-    public void testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided() throws Exception {
+    public void testUpdateTrainsToUndefinedIfNotEnoughDataWasProvided()  {
         final Bridge bridge = mockBridge();
         final ThingHandlerCallback callback = mock(ThingHandlerCallback.class);
 
@@ -179,8 +186,8 @@ public class DeutscheBahnTimetableHandlerTest implements TimetablesV1ImplTestHel
                     argThat(arg -> arg.getStatus().equals(ThingStatus.ONLINE)));
 
             verifyThingUpdated(bridge, 0, stop01.getId());
-            verifyChannelsUpdatedToUndef(bridge, 1, callback, UnDefType.UNDEF);
-            verifyChannelsUpdatedToUndef(bridge, 2, callback, UnDefType.UNDEF);
+            verifyChannelsUpdatedToUndef(bridge, 1, callback);
+            verifyChannelsUpdatedToUndef(bridge, 2, callback);
 
         } finally {
             handler.dispose();

